exploration/quoted-literals.md

```diff
@@ -0,0 +1,390 @@
+  > -- This is impossible in MessageFormat 2.0.
+  > {+button title=|Goodbye, {$userName}!|}Sign out{-button}
+  > ```
+
+- Selector function implementers may want to support multi-word variant keys
+  or exotic characters in variant keys
+  to effectively create "mini-DSLs" for the matching logic:
+
+  > ```
+  > {{ match {$count :choice}
+  >    when |<10| {{A handful.}}
+  >    when |11..19| {{Umpteen.}}
+  >    when * {{Lots.}}
+  > }}
+  >
+  > {{ match {$arbitraryString}
+  >    when |can't resolve| {{Can't resolve!}}
+  >    when |11'233.44| {{Locale formatted number}}
+  >    when |New York| {{A multi-word proper name}}
+  >    when * {{Imagine more...}}
+  > }}
+  > ```
+
+- Message authors may want to protect untranslatable strings:
+
+  > ```
+  > Visit {|http://www.example.com| @translate=false}.
+  > ```
+  >
+  > See the [expression attributes design proposal](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Message authors may want to decorate substrings as being written in a particular language,
+  different from the message's language,
+  for the purpose of accessibility, text-to-speech, and semantic correctness.
+
+  > ```
+  > The official native name of the Republic of Poland is {|Rzeczpospolita Polska| @lang=pl}.
+  > ```
+  >
+  > See the [expression attributes design proposal](https://github.com/unicode-org/message-format-wg/blob/main/exploration/0002-expression-attributes.md).
+
+- Developers may want to embed messages with quoted literals in code written in another programming language
+  which uses single or double quotes to delimit strings.
+
+  > ```js
+  > let message = new MessageFormat('en', 'A message with {|a literal|}.');
+  > ```
+
+- Developers and localization engineers may want to embed messages with quoted literals in a container format, such as JSON.
+
+  > ```json
+  > {
+  >   "msg": "A message with {|a literal|}."
+  > }
+  > ```
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- **[r1; high priority]** Minimize the need to escape characters inside literals.
+  In particular, choose a delimiter that isn't frequently used in translation content.
+  Having to escape characters inside literals is inconvenient and error-prone when done by hand,
+  and it also introduces the backslash into the message, `\`,
+  which is the escape introducer.
+  The backslash then needs to be escaped too,
+  when the message is embedded in code or containers.
+  (This is how some syntaxes produce the gnarly `\\\`.)
+
+- **[r2; high priority]** Minimize the need to escape characters when embedding messages in code or containers.
+  In particular, choose a delimiter that isn't frequently used as a string delimiter in programming languages and container formats.
+  However, note that many programming languages also provide alternative ways of delimiting strings, e.g. _raw strings_ or triple-quoted literals.
+
```

Comment:
I think we're missing a requirement? We have minimization of escaping inside the literal and minimization of escaping for the host format. We also don't want to choose a character/sequence that requires normal patterns to have lots of escapes.

```suggestion
- **[r2a; high priority]** Minimize the need to escape characters in patterns when writing normal messages.
```

What I mean here is:
```
{{This has \|an escape\| because it needs it}}
{{This has \'an escape\' because it needs it}}
```

Reply:
Why would these patterns require any escaping?